File: 76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ""
        
        countT = {}
        window = {}

        for c in t:
            countT[c] = 1 + countT.get(c , 0)
        
        l = 0 
        res = ""
        reslen = float('inf')
        need = len(set(t))
        have = 0

        for r in range(len(s)):
            window[s[r]] = 1 + window.get(s[r] , 0)

            c = s[r]

            if c in countT and window[c] == countT[c]:
                have += 1
            
            while have == need:

                if (r - l + 1) < reslen:
                    reslen = r - l + 1
                    res = s[l : r + 1]

                
                window[s[l]] -= 1

                if s[l] in countT and window[s[l]] < countT[s[l]]:
                    have -= 1

                l += 1
        return res
                

                

File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {

        //INITAIL EDGE CASE 

        if(s.length() < t.length())
        {
            return "";
        }

        HashMap<Character , Integer> countT = new HashMap<>();
        HashMap<Character , Integer> window = new HashMap<>();

        for(int i = 0; i < t.length() ; i++)
        {
            char ch = t.charAt(i);
            countT.put(ch , countT.getOrDefault(ch , 0) + 1);
        }

        String res = "";
        int reslen = Integer.MAX_VALUE;
        int l = 0;
        int need = countT.size();
        int have = 0;

        for(int r = 0 ; r < s.length(); r++)
        {
            char c = s.charAt(r);
            window.put(c , window.getOrDefault(c , 0) + 1);

            if(countT.containsKey(c) && countT.get(c).equals(window.get(c)))
            {
                have++;
            }

            while(have == need)
            {
                if((r - l + 1) < reslen)
                {
                    reslen = r - l + 1;
                    res = s.substring(l , r + 1);

                }
                char ch = s.charAt(l);
                window.put(ch , window.get(ch) - 1);

                if(countT.containsKey(ch) && countT.get(ch) > window.get(ch))
                {
                    have--;
                }

                l++;
            }

        }
        return res;

        
    }
}